cache/lib/cache_web/controllers/module_cache_controller.ex

```diff
@@ -0,0 +1,331 @@
+      key = Disk.module_key(account_handle, project_handle, category, hash, name)
+
+      if Disk.module_exists?(account_handle, project_handle, category, hash, name) or S3.exists?(key) do
+        send_resp(conn, :no_content, "")
+      else
+        conn
+        |> put_status(:not_found)
+        |> json(%{message: "Artifact not found"})
+      end
+    else
+      {:error, message} ->
+        conn
+        |> put_status(:bad_request)
+        |> json(%{message: message})
+    end
+  end
+
+  defp parse_project_id(%{project_id: project_id}) when is_binary(project_id) do
+    case String.split(project_id, "/") do
+      [account_handle, project_handle] -> {:ok, {account_handle, project_handle}}
+      _ -> {:error, "Invalid project_id format. Expected 'account_handle/project_handle'"}
+    end
+  end
+
+  defp parse_project_id(_), do: {:error, "Missing required parameter: project_id"}
+
+  defp fetch_param(params, key) do
+    case Map.get(params, key) do
+      nil -> {:error, "Missing required parameter: #{key}"}
+      value -> {:ok, value}
+    end
+  end
+
+  defp handle_existing_artifact(conn) do
+    :telemetry.execute([:cache, :module, :upload, :exists], %{count: 1}, %{})
+
+    case BodyReader.drain(conn) do
+      {:ok, conn_after} -> send_resp(conn_after, :no_content, "")
+      {:error, conn_after} -> send_resp(conn_after, :no_content, "")
+    end
+  end
+
+  defp save_new_artifact(conn, account_handle, project_handle, category, hash, name) do
+    case BodyReader.read(conn) do
+      {:ok, data, conn_after} ->
+        size = get_data_size(data)
+        :telemetry.execute([:cache, :module, :upload, :attempt], %{size: size}, %{})
+        persist_artifact(conn_after, account_handle, project_handle, category, hash, name, data, size)
+
+      {:error, :too_large, conn_after} ->
+        :telemetry.execute([:cache, :module, :upload, :error], %{count: 1}, %{reason: :too_large})
+        send_error(conn_after, :request_entity_too_large, "Request body exceeded allowed size")
+
+      {:error, :timeout, conn_after} ->
+        :telemetry.execute([:cache, :module, :upload, :error], %{count: 1}, %{reason: :timeout})
+        send_error(conn_after, :request_timeout, "Request body read timed out")
+
+      {:error, :cancelled, conn_after} ->
+        :telemetry.execute([:cache, :module, :upload, :cancelled], %{count: 1}, %{})
+        send_resp(conn_after, :no_content, "")
+
+      {:error, _reason, conn_after} ->
+        :telemetry.execute([:cache, :module, :upload, :error], %{count: 1}, %{reason: :read_error})
+        send_error(conn_after, :internal_server_error, "Failed to persist artifact")
+    end
+  end
+
+  defp get_data_size({:file, tmp_path}) do
+    case File.stat(tmp_path) do
+      {:ok, %File.Stat{size: sz}} -> sz
+      _ -> 0
+    end
+  end
+
+  defp get_data_size(bin) when is_binary(bin), do: byte_size(bin)
+
+  defp persist_artifact(conn, account_handle, project_handle, category, hash, name, data, size) do
```

Comment:
I'd recommend to have simple or no business logic in the controllers beyond validation, authorization, and mapping the result of some business logic to the JSON response. This controller seems to be doing a lot.